src/lib.rs

```diff
@@ -198,7 +222,46 @@ impl Contract {
         notify::notify_mentions(desc.as_str(), id);
     }
 
+    #[payable]
+    pub fn add_proposal(&mut self, body: VersionedProposalBody, labels: HashSet<String>) {
+        near_sdk::log!("add_proposal");
+        let id = self.proposals.len();
+        let author_id = env::predecessor_account_id();
+        let editor_id = author_id.clone();
+        require!(
+            self.is_allowed_to_use_labels(
+                Some(editor_id.clone()),
+                labels.iter().cloned().collect()
+            ),
+            "Cannot use these labels"
+        );
+
+        for label in &labels {
+            let mut other_proposals = self.label_to_proposals.get(label).unwrap_or_default();
+            other_proposals.insert(id);
+            self.label_to_proposals.insert(label, &other_proposals);
+        }
+        let proposal = Proposal {
+            id,
+            author_id: author_id.clone(),
+            block_height: 0u64, // TODO
+            snapshot: ProposalSnapshot { editor_id, timestamp: env::block_timestamp(), labels, body },
+            snapshot_history: vec![],
+        };
+        self.proposals.push(&proposal.clone().into());
+
+        let mut author_proposals = self.author_proposals.get(&author_id).unwrap_or_else(|| HashSet::new());
+        author_proposals.insert(proposal.id);
+        self.author_proposals.insert(&proposal.author_id, &author_proposals);
+
+        let desc = get_proposal_description(proposal.clone());
+
+        proposal::repost::publish_to_socialdb_feed(proposal);
```

Comment:
You need to chain a [callback](https://docs.near.org/sdk/rust/cross-contract/callbacks) here to capture the block height received from SocialDB.